### PR TITLE
*: implement System.Blockchain.GetTransaction and System.Blockchain.GetBlock interops

### LINKED ISSUE #1023: Implement new System.Blockchain.GetTransaction interop
It replaces old `GetTransaction` as well as lots of other related interops (that are no longer needed).

### PR DESCRIPTION
**Pull request by AnnaShaleva**
closes #1023 and #1025

### CODE REVIEW

File: pkg/core/interop_system.go [hunk 1]
```diff
@@ -112,13 +115,27 @@ func getTransactionAndHeight(cd *dao.Cached, v *vm.VM) (*transaction.Transaction
...
 }
 
+// isTraceableBlock defines whether we're able to give information about
+// the block with index specified.
+func isTraceableBlock(ic *interop.Context, index uint32) bool {
+	height := ic.Chain.BlockHeight()
+	if index > height {
+		return false
+	}
+	return index+MaxTraceableBlocks > height
```
Comment (fyrchik) [inline pkg/core/interop_system.go:125]: `return index <= height && index + MaxTraceableBlocks > height` ?

File: pkg/interop/block/block.go [hunk 2]
```diff
@@ -19,12 +19,12 @@ func GetTransactionCount(b Block) int {
 
 // GetTransactions returns a slice of transactions recorded in the given block.
 // It uses `Neo.Block.GetTransactions` syscall internally.
-func GetTransactions(b Block) []transaction.Transaction {
-	return []transaction.Transaction{}
+func GetTransactions(b Block) []blockchain.Transaction {
+	return []blockchain.Transaction{}
```
Comment (fyrchik) [inline pkg/interop/block/block.go]: Implementation of this and next interop should also be changed.

File: pkg/core/interop_system.go [hunk 3]
```diff
@@ -145,6 +145,10 @@ func bcGetTransactionHeight(ic *interop.Context, v *vm.VM) error {
 	if err != nil {
 		return err
```
Comment (fyrchik) [inline pkg/core/interop_system.go:166]: Should we return `-1` here now?

File: pkg/core/interop_system.go [hunk 4]
```diff
@@ -139,6 +139,26 @@ func bcGetTransaction(ic *interop.Context, v *vm.VM) error {
...
+// bcGetTransactionFromBlock returns transaction with the given index from the
+// block with height or hash specified.
+func bcGetTransactionFromBlock(ic *interop.Context, v *vm.VM) error {
+	hash, err := getBlockHashFromElement(ic.Chain, v.Estack().Pop())
+	if err != nil {
+		return err
+	}
+	block, err := ic.Chain.GetBlock(hash)
+	if err != nil || !isTraceableBlock(ic, block.Index) {
+		v.Estack().PushVal(stackitem.Null{})
```
Comment (fyrchik) [inline pkg/core/interop_system.go:151]: We should return after this.

File: pkg/core/interop_system.go [hunk 5]
```diff
@@ -139,6 +139,26 @@ func bcGetTransaction(ic *interop.Context, v *vm.VM) error {
...
+	}
+	index := v.Estack().Pop().BigInt().Int64()
+	if index < 0 || index >= int64(len(block.Transactions)) {
```
Comment (fyrchik) [inline pkg/core/interop_system.go:154]: C# implementation uses `len-1` (and `index+1` below). Probably 1st tx shouldn't be taken into account.

File: pkg/interop/block/block.go (hunk 2 above)
Comment (fyrchik) [reply pkg/interop/block/block.go]: Oh, I see they were remove later. Probably worth removing before.

File: pkg/core/block/block.go [hunk 6]
```diff
@@ -235,3 +237,18 @@ func (b *Block) UnmarshalJSON(data []byte) error {
 	_ = b.ConsensusData.Hash()
 	return nil
 }
+
+// ToStackItem converts Block to stackitem.Array
+func (b *Block) ToStackItem() stackitem.Item {
```
Comment (roman-khimov) [inline pkg/core/block/block.go:242]: Same thing here, let's not create a dependency on `stackitem` until we really need it.

File: pkg/interop/blockchain/blockchain.go [hunk 7]
```diff
@@ -35,6 +34,32 @@ type Transaction struct {
 	Scrypt []byte
 }
 
+// Block represents a NEO block, it's an opaque data structure that you can get
```
Comment (roman-khimov) [inline pkg/interop/blockchain/blockchain.go:37]: Not opaque.

File: pkg/core/interop_system.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/core/interop_system.go]: Changed to -1.

File: pkg/core/interop_system.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/core/interop_system.go:178]: That's true.

File: pkg/compiler/codegen.go [hunk 8]
```diff
@@ -1051,6 +1051,10 @@ func (c *codegen) convertSyscall(expr *ast.CallExpr, api, name string) {
 		emit.Opcode(c.prog.BinWriter, opcode.PACK)
 	}
 	emit.Syscall(c.prog.BinWriter, api)
+	switch name {
+	case "GetTransaction":
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/compiler/codegen.go]: Do we really need to use the full path to the package? We check this in `isSyscall` function before `convertSyscall` call. Or did you mean to check `api` only (like `GetTransaction` belongs to `blockchain` package)? In this case after we complete all interops adjustment, there will remain the only `GetTransaction` method.

File: pkg/core/transaction/transaction.go [hunk 9]
```diff
@@ -307,3 +309,17 @@ func (t *Transaction) UnmarshalJSON(data []byte) error {
 
 	return nil
 }
+
+// ToStackItem converts transaction to stackitem.Item
+func (t *Transaction) ToStackItem() stackitem.Item {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/transaction/transaction.go]: Moved this method to `interop_system.go`.

File: pkg/interop/blockchain/blockchain.go [hunk 10]
```diff
@@ -8,9 +8,33 @@ import (
...
 )
 
+// Transaction represents a NEO transaction, it's an opaque data structure
+// that can be used with functions from this package. It's similar to
+// Transaction class in Neo .net framework.
+type Transaction struct {
+	// Hash represents the hash (256 bit BE value in a 32 byte slice) of the
+	// given transaction (which also is its ID).
+	Hash []byte
+	// Version represents the trading version.
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/interop/blockchain/blockchain.go]: Hm, just copied this from `transaction.Transaction`. Replaced to `transaction version`. BTW, what's the real meaning of this field?

File: pkg/core/interop_system.go [hunk 11]
```diff
@@ -139,6 +139,26 @@ func bcGetTransaction(ic *interop.Context, v *vm.VM) error {
...
 }
 
+// bcGetTransactionFromBlock returns transaction with the given index from the
+// block with height or hash specified.
+func bcGetTransactionFromBlock(ic *interop.Context, v *vm.VM) error {
+	hash, err := getBlockHashFromElement(ic.Chain, v.Estack().Pop())
+	if err != nil {
+		return err
+	}
+	block, err := ic.Chain.GetBlock(hash)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/interop_system.go]: Replaced it with `DAO.GetBlock`.

File: pkg/compiler/codegen.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/compiler/codegen.go]: OK, I think the check in the `isSyscall()` is enough.

File: pkg/interop/blockchain/blockchain.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/interop/blockchain/blockchain.go]: Well, it defines a transaction format version. If we're to add/remove some fields we'll bump this version to a newer one. Neo 2 had done it actually once (you can find some branches in the code where transaction processing differs for version 0 and version 1).

**Pull request merged**